sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala

```diff
@@ -279,6 +283,16 @@ class ArtifactManager(session: SparkSession) extends Logging {
     loader
   }
 
+  private[sql] def clone(newSession: SparkSession): ArtifactManager = {
+    val newArtifactManager = new ArtifactManager(newSession)
+    if (artifactPath.toFile.exists()) {
+      FileUtils.copyDirectory(artifactPath.toFile, newArtifactManager.artifactPath.toFile)
+    }
+    newArtifactManager.jarsList.addAll(jarsList)
+    newArtifactManager.pythonIncludeList.addAll(pythonIncludeList)
+    newArtifactManager
```

Comment:
We'd still need to handle cached relations because of the way the key is formed:
```
        val blockId = CacheId(
          sessionUUID = session.sessionUUID,
          hash = normalizedRemoteRelativePath.toString.stripPrefix(s"cache${File.separator}"))
```
The `session.sessionUUID` would differ in the cloned session (because the `SparkSession` is a new instance)

To handle this, we can simply re-insert the blocks in the BlockManager with the new `blockId` but we'd need to store all of the currently registered blobs like the jars/python includes in order to do this